pkg/util/restoptions/configgetter.go

```diff
@@ -0,0 +1,139 @@
+// By default, the etcd watch cache is enabled with a size of 1000 per resource type.
+func NewConfigGetter(masterOptions configapi.MasterConfig) Getter {
+	getter := &configRESTOptionsGetter{
+		masterOptions:    masterOptions,
+		cacheEnabled:     true,
+		defaultCacheSize: 1000,
+		cacheSizes:       map[unversioned.GroupResource]int{},
+		restOptionsMap:   map[unversioned.GroupResource]genericrest.RESTOptions{},
+	}
+
+	if err := getter.loadWatchCacheSettings(); err != nil {
+		glog.Error(err)
+	}
+
+	return getter
+}
+
+func (g *configRESTOptionsGetter) loadWatchCacheSettings() error {
+	if g.masterOptions.KubernetesMasterConfig == nil {
+		return nil
+	}
+
+	server := apiserveroptions.NewAPIServer()
+	if errs := cmdflags.Resolve(g.masterOptions.KubernetesMasterConfig.APIServerArguments, server.AddFlags); len(errs) > 0 {
+		return kerrors.NewAggregate(errs)
+	}
+
+	g.cacheEnabled = server.EnableWatchCache
+
+	errs := []error{}
+	for _, c := range server.WatchCacheSizes {
+		tokens := strings.Split(c, "#")
+		if len(tokens) != 2 {
+			errs = append(errs, fmt.Errorf("invalid watch cache size value '%s', expecting <resource>#<size> format (e.g. builds#100)", c))
+			continue
+		}
+
+		resource := unversioned.ParseGroupResource(tokens[0])
+
+		size, err := strconv.Atoi(tokens[1])
+		if err != nil {
+			errs = append(errs, fmt.Errorf("invalid watch cache size value '%s': %v", c, err))
+			continue
+		}
+
+		g.cacheSizes[resource] = size
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+func (g *configRESTOptionsGetter) GetRESTOptions(resource unversioned.GroupResource) (genericrest.RESTOptions, error) {
+	g.restOptionsLock.Lock()
+	defer g.restOptionsLock.Unlock()
+	if resourceOptions, ok := g.restOptionsMap[resource]; ok {
+		return resourceOptions, nil
+	}
+
+	if g.etcdHelper == nil {
+		// TODO: choose destination etcd based on input resource
+		etcdClient, err := etcd.MakeNewEtcdClient(g.masterOptions.EtcdClientInfo)
+		if err != nil {
+			return genericrest.RESTOptions{}, err
+		}
+		// TODO: choose destination group/version based on input group/resource
+		groupVersion := unversioned.GroupVersion{Group: "", Version: g.masterOptions.EtcdStorageConfig.OpenShiftStorageVersion}
+		g.etcdHelper = etcdstorage.NewEtcdStorage(etcdClient, kapi.Codecs.LegacyCodec(groupVersion), g.masterOptions.EtcdStorageConfig.OpenShiftStoragePrefix, false)
+	}
+
+	decorator := genericrest.UndecoratedStorage
+	if g.cacheEnabled {
+		configuredSize, specified := g.cacheSizes[resource]
+		if !specified || configuredSize <= 0 {
+			configuredSize = g.defaultCacheSize
+		}
+		glog.V(5).Infof("using watch cache size of %d for %s", configuredSize, resource.String())
+
+		decorator = func(s storage.Interface, requestedSize int, objectType runtime.Object, resourcePrefix string, scopeStrategy rest.NamespaceScopedStrategy, newListFunc func() runtime.Object) storage.Interface {
+			capacity := requestedSize
+			if capacity == UseDefaultCacheSize {
```

Comment:
What a strange API.  Specifying the "UseDefault" value actually results in "UseConfigured", not "use default".  If I've actually configured the value, there's no way to say "use default", right?

Also, this means that if I set `0` in hopes of not caching for a particular resource and we end up supporting it, the `requestedSize` (actually an `overrideSize`?) won't do anything?  I'm probably ok with that.

Why are there so many layers?